core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -874,7 +874,13 @@ class Partition(val topicPartition: TopicPartition,
   private def createLogInAssignedDirectoryId(partitionState: LeaderAndIsrPartitionState, highWatermarkCheckpoints: OffsetCheckpoints, topicId: Option[Uuid], targetLogDirectoryId: Option[Uuid]): Unit = {
     targetLogDirectoryId match {
       case Some(directoryId) =>
-        createLogIfNotExists(directoryId == DirectoryId.UNASSIGNED, isFutureReplica = false, highWatermarkCheckpoints, topicId, targetLogDirectoryId)
+        if (logManager.onlineLogDirId(directoryId) || !logManager.hasOfflineLogDirs() || directoryId == DirectoryId.UNASSIGNED) {
+          createLogIfNotExists(partitionState.isNew, isFutureReplica = false, highWatermarkCheckpoints, topicId, targetLogDirectoryId)
+        } else {
+          warn(s"Skipping creation of log because there are potentially offline log " +
```

Comment:
I'm not quite sure if we should also be throwing an exception here.

Reply:
TBH, I don't understand why we should throw exception [here](https://github.com/apache/kafka/blob/f1924353126fdf6aad2ba1f8d0c22dade59360b1/core/src/main/scala/kafka/log/LogManager.scala#L1009). Is it because there are potentially offline log dir might already have the logs? If so, does that mean we can't create any new topic while 1 log dir goes offline? Sorry, I'm not quite familiar with this part of logic.

Reply:
IIUC, it's from [KIP-112](https://cwiki.apache.org/confluence/display/KAFKA/KIP-112%3A+Handle+disk+failure+for+JBOD)

> The broker will specify error=KafkaStorageException for those partitions that are in the LeaderAndIsrRequest with isNewReplica=False but not found on any good log directory.

Reply:
> Is it because there are potentially offline log dir might already have the logs? If so, does that mean we can't create any new topic while 1 log dir goes offline? 

My understanding of the original logic before JBOD is that we block creating topic partition on broker if `partitionState.isNew` is false and the broker has any offline log directories. This logic is in `LogManager.getOrCreateLog` which will throw `KafkaStorageException`. Am not sure if we need to throw it here in `Partition.createLogInAssignedDirectoryId` if it will be thrown anyway in  `LogManager.getOrCreateLog`

Reply:
Yes, I agree with @OmniaGM  that we don't need to branch if/else here since we already do the handling in `createLogIfNotExists`. So we only need to pass in `partitionState.isNew` correctly here. Does that make sense?

Reply:
Also, before this change, we'll always treat it as `new` if `directoryId == DirectoryId.UNASSIGNED`. But now, it needs to be `directoryId == DirectoryId.UNASSIGNED && partitionState.isNew`. Will that cause other issue @OmniaGM ?

Reply:
@OmniaGM , thanks for the suggestion. @gaurav-narula , could you please help add more tests for the different cases?

Reply:
Addressed in [9678cc9](https://github.com/apache/kafka/pull/15263/commits/9678cc9c19b322232a78e2623bbada6b51e55baa)

Reply:
What is the operator supposed to do when they see this warning? Generally, we should be very careful about warning logs - most times they are an anti-pattern as they are scary but not actionable.